streams/js/src/main/scala/zio/stream/platform.scala

```diff
@@ -2,12 +2,60 @@ package zio.stream
 
 import java.io.{ IOException, InputStream }
 
+import scala.concurrent.Future
+
 import zio._
 
 trait ZSinkPlatformSpecificConstructors
 
 trait ZStreamPlatformSpecificConstructors { self: ZStream.type =>
 
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The optionality of the error type `E` can be used to signal the end of the stream,
+   * by setting it to `None`.
+   */
+  def effectAsync[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => Unit,
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    effectAsyncMaybe(callback => {
+      register(callback)
+      None
+    }, outputBuffer)
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The registration of the callback can possibly return the stream synchronously.
+   * The optionality of the error type `E` can be used to signal the end of the stream,
+   * by setting it to `None`.
+   */
+  def effectAsyncMaybe[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => Option[ZStream[R, E, A]],
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    ZStream {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        maybeStream <- ZManaged.effectTotal {
+                        register(k => runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer)))
```

Comment:
@iravid it's being used actually. In the test cases it's used to compose futures and guarantee sequentially. 

It wasn't returning future at some point and test were failing because it didn't wait for previous `callbacks` to finish.